Internal/narrow types required for compilation

cqc-melf opened the pull request:
# Description

Please summarise the changes.

# Related issues

Please mention any github issues addressed by this PR.

# Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented hard-to-understand parts of my code.
- [ ] I have made corresponding changes to the public API documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the changelog with any user-facing changes.

mwpb commented:
Thanks for inviting me to the repo. The test failures were because I didn't appreciate that `BackendProperties` was optional. Now that has been corrected the tests pass and I was able to clean up a few instances of `cast`. Some remaining questions from me:-
* Can we think of a better name than `default_compilation_pass_static`? The trouble with `default_compilation_pass_from_config` is that it is already a little inaccurate and may get more inaccurate in the future. An alternative might be `default_compilation_pass_offline`.
* Should I document the new method in `docs/api.rst` ? I don't know what additional uses outside of Tierkreis it would have.

Otherwise, I'll simply look forward to your review and take care of any issues arising.

cqc-melf commented on docs/changelog.rst:
```diff
@@ -6,6 +6,7 @@ Unreleased
 
 * Update pytket version requirement to 1.26.
 * Update qiskit-aer version requirement to 0.14.
+* Add `IBMQBackend.default_compilation_pass_static` for offline compilation given config and props objects.
```
I would suggest to put `offline` in the name somewhere, maybe just do ``IBMQBackend.default_compilation_pass_offline` ?

mwpb replied:
Sounds good. Done.

cqc-melf commented on pytket/extensions/qiskit/backends/ibm.py:
```diff
@@ -51,15 +51,20 @@
 
 from pytket.circuit import Circuit, OpType
-from pytket.backends import Backend, CircuitNotRunError, CircuitStatus, ResultHandle
+from pytket.backends import (
```
Are you using a different lint / format?

mwpb replied:
Nope. Just using Black. I removed the trailing comma and Black put it back to its original form!

cqc-melf commented on pytket/extensions/qiskit/backends/ibm.py:
```diff
@@ -239,9 +247,12 @@ def backend_info(self) -> BackendInfo:
+        cls,
+        config: QasmBackendConfiguration,
+        props: Optional[BackendProperties],
+    ) -> BackendInfo:
```
I know that this is not a public function, but could you maybe add a short comment on what this function is doing?

mwpb replied:
Sure. Done.

cqc-melf commented on pytket/extensions/qiskit/backends/ibm.py:
```diff
@@ -310,9 +321,12 @@ def available_devices(cls, **kwargs: Any) -> List[BackendInfo]:
+        backend_info_list = []
+        for backend in provider.backends():
+            config = backend.configuration()
+            props = backend.properties()
```
Have you checked if the properties you need at this point are not in the config?

mwpb replied:
Yes I did try initially to just use the config. My reasoning that the properties should be passed also went as follows... `_get_backend_info` calls `process_characterisation_from_config`, which does use the properties if available. The currently published code (before this PR) does seem to avail itself of properties where it exists because the first thing that `process_characterisation` does is call
```properties = cast("BackendProperties", backend.properties())```

cqc-melf replied:
Ah okay! Sounds good

cqc-melf commented on pytket/extensions/qiskit/backends/ibm.py:
```diff
@@ -376,40 +390,57 @@ def default_compilation_pass(
+        )
+
+    @staticmethod
+    def default_compilation_pass_static(
```
maybe rename this as well to `...offline` ?

mwpb replied:
Done. Also changed `rebase_pass_static` to `rebase_pass_offline` for consistency.

cqc-melf reviewed (commented):
Hi @mwpb ,
I have written a few comments and questions, if there is anything you want to discuss in a call, please let me know :)

mwpb commented:
Thanks for the comments Melf! I'll work through them either today or tomorrow.

mwpb commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -873,8 +870,24 @@ def process_characterisation(backend: "QiskitBackend") -> Dict[str, Any]:
+    :rtype: dict
+    """
+
+    # TODO explicitly check for and separate 1 and 2 qubit gates
```
Ah thanks, have corrected the docstring and removed the extra TODO comment. The TODO itself was there already and I've kept it in its right place.

mwpb commented:
@cqc-melf thanks for the comments. Some initial responses above.

mwpb commented:
Just to clarify, @cqc-melf  this should be ready for your review when you have time.

cqc-melf commented:
@mwpb Thank you for the update and addressing all the comments.
I think there is no need to add this to the public documentation.

There is one conflict with develop, can you maybe resolve that?

Besides that, I am happy to approve this :)

cqc-melf commented:
Looks good, but I can't approve this as this is my own PR. Can you? @mwpb

cqc-melf commented:
Great :) Please do a squash and merge
